src/use-observable.ts

```diff
@@ -2,42 +2,52 @@ import { Observable, BehaviorSubject } from 'rxjs'
 import { useState, useEffect, useMemo } from 'react'
 
 export type InputFactory<T, U = undefined> = U extends undefined
-  ? () => Observable<T>
-  : (props$: Observable<U>) => Observable<T>
+  ? (state$: Observable<T>) => Observable<T>
+  : (props$: Observable<U>, state$: Observable<T>) => Observable<T>
 
 export function useObservable<T>(inputFactory: InputFactory<T>): T | null
 export function useObservable<T>(inputFactory: InputFactory<T>, initialState: T): T
-export function useObservable<T, U extends ReadonlyArray<any>>(
-  inputFactory: InputFactory<T, U>,
-  initialState: T,
-  inputs: U,
-): T
+export function useObservable<T, U>(inputFactory: InputFactory<T, U>, initialState: T, inputs: U): T
 
-export function useObservable<T, U extends ReadonlyArray<any> | undefined>(
-  inputFactory: InputFactory<T, U>,
-  initialState?: T,
-  inputs?: U,
-): T | null {
-  const [inputs$] = useState(new BehaviorSubject<U | undefined>(inputs))
-  const [state, setState] = useState(typeof initialState !== 'undefined' ? initialState : null)
+export function useObservable<T, U>(inputFactory: InputFactory<T, U>, initialState?: T, inputs?: U): T | null {
+  const stateSubject$ = new BehaviorSubject<T | undefined>(initialState)
+  const inputSubject$ = new BehaviorSubject<U | undefined>(inputs)
+  const [[state, state$, inputs$], setState] = useState<
+    [T | null, BehaviorSubject<T | undefined>, BehaviorSubject<U | undefined>]
+  >([typeof initialState !== 'undefined' ? initialState : null, stateSubject$, inputSubject$])
```

Comment:
I am not sure if we should use three `useState` to do the same thing here? And I guess we don't need `setState$` or `setInputs$`.

refer: https://reactjs.org/docs/hooks-faq.html#should-i-use-one-or-many-state-variables